BookingsApi/BookingsApi.DAL/Services/HearingAllocationService.cs

```diff
@@ -0,0 +1,204 @@
+            if (hearing == null)
+            {
+                throw new DomainRuleException("HearingNotFound",
+                    $"Hearing {hearingId} not found");
+            }
+
+            var allocations = _context.Allocations.Where(a => a.HearingId == hearing.Id).ToList();
+            if (allocations.Any())
+            {
+                throw new DomainRuleException("HearingAlreadyAllocated",
+                    $"Hearing {hearing.Id} has already been allocated");
+            }
+
+            var hearingStartTime = hearing.ScheduledDateTime;
+            var hearingEndTime = hearing.ScheduledEndTime;
+
+            if (hearingStartTime.Date != hearingEndTime.Date)
+            {
+                throw new DomainRuleException("AllocationNotSupported",
+                    $"Unable to allocate to hearing {hearing.Id}, hearings which span multiple days are not currently supported");
+            }
+
+            var cso = SelectCso(hearingStartTime, hearingEndTime);
+            if (cso == null)
+            {
+                throw new DomainRuleException("NoCsosAvailable",
+                    $"Unable to allocate to hearing {hearingId}, no CSOs available");
+            }
+
+            return cso;
+        }
+
+        private JusticeUser SelectCso(DateTime hearingStartTime, DateTime hearingEndTime)
+        {
+            var availableCsos = GetAvailableCsos(hearingStartTime, hearingEndTime);
+
+            if (!availableCsos.Any())
+            {
+                return null;
+            }
+
+            if (availableCsos.Count() == 1)
+            {
+                return availableCsos.SingleOrDefault();
+            }
+
+            var csos = availableCsos
+                .Select(c => new
+                {
+                    Cso = c,
+                    AllocationCount = c.Allocations?.Count ?? 0
+                })
+                .ToList();
+            
+            var lowestAllocationCount = csos.Min(c => c.AllocationCount);
+            var csosWithFewestAllocations = csos.Where(c => c.AllocationCount == lowestAllocationCount).ToList();
+
+            if (csosWithFewestAllocations.Count == 1)
+            {
+                return csosWithFewestAllocations.Single().Cso;
+            }
+
+            return SelectRandomly(csosWithFewestAllocations.Select(c => c.Cso).ToList());
+        }
+
+        private IEnumerable<JusticeUser> GetAvailableCsos(DateTime hearingStartTime, DateTime hearingEndTime)
+        {
+            var availableCsos = new List<JusticeUser>();
+
+            var csos = _context.JusticeUsers
+                .Include(u => u.UserRole)
+                .Where(u => u.UserRoleId == (int)UserRoleId.Vho)
+                .ToList();
+
+            var csoUserIds = csos.Select(u => u.Id).ToList();
+
+            var workHours = _context.VhoWorkHours
+                .Where(wh => csoUserIds.Contains(wh.JusticeUserId))
+                .AsEnumerable()
```

Comment:
is the AsEnumerable() needed here?

Reply:
I was using ToList() originally, but SonarCloud suggested to replace it with AsEnumerable().

One of them is needed in order to pull the data into memory so that I can then use the SystemDayOfWeek calculated property (which can't be translated into SQL).

Reply:
Ok cool. That makes sense